### DISCUSSION TITLE
Problem with accessing response from callback

### DISCUSSION THREAD

**Discussion by Matys-check** [mockoon/mockoon#1345] [Help] [answered]
Hi, I need to create a mock for an asynchronous service with callback. Everything would be fine if not for the problem I encountered namely
In the response that the mock generates I have the generation of correlationId, which I then need to send via callback in order to map the queries.

My response definition:
`{
  "expTime": "2020-05-05",
  "replayTo": "{{body 'element.0.replayTo' 'default'}}",
  "correlationId": "{{ faker 'string.uuid' }}"
}`

And my callback definition:
`{
"name": "It's callback",  
"correlationId": "{{ response.correlationId }}",
"returnUrl": "{{body 'element.0.replayTo' 'default'}}",
"log": "{{ response }}"
}`

I have tried many solutions unfortunately I have no idea how from the callback level I can get to the variables in the response.
The pasted solutions generates empty values, the response object is also empty.

![image](https://github.com/mockoon/mockoon/assets/84642745/aa1949f9-c7c0-4c9e-ac71-0caef696169a)

**Comment by 255kb** [comment]
Hi, 

There is currently no helper allowing access to the response properties.
What you could try is to set a [global variable](https://mockoon.com/docs/latest/variables/global-variables/) in the response and get it in the callback. However, due to a [bug (already fixed)](https://github.com/mockoon/mockoon/issues/1300) this will only be available in the next release. 
But you will be able to do something like:

```
{{setGlobalVar 'correlationId' (faker 'string.uuid')}}
{ "expTime": "2020-05-05", "replayTo": "{{body 'element.0.replayTo' 'default'}}", "correlationId": "{{ getGlobalVar 'correlationId' }}" }
```

```
{ "name": "It's callback",   "correlationId": "{{ getGlobalVar 'correlationId' }}", "returnUrl": "{{body 'element.0.replayTo' 'default'}}", "log": "{{ response }}" }
```

But a series of `response` helpers would be useful, indeed. As always, contributions are welcomed :)

**Comment by Matys-check** [comment]
Hello @255kb , sorry for the delay. I can finally return to the topic, and indeed, the solution you provided is almost correct. I say "almost" because:
If we take a closer look at the values of our "correlationId," we can see that:

Response:
"correlationId": "8c134edc-3b50-4360-a86e-b7cb20a1c906"
Callback:
"correlationId": "ef1723ac-638c-4cc5-95ef-a400e6d8ac51"

This suggests that referencing the variable correlationId causes it to be regenerated, leading to a mismatch.
I also tried creating an intermediate variable to avoid directly referencing the correlationId variable, but unfortunately, that didn’t help either. A potentially helpful detail is that I have a 1-second delay on the response and a 5-second delay on the callback.
Response definition:
`{{setGlobalVar 'correlationId' (faker 'string.uuid')}}
{{setGlobalVar 'anotherId' (getGlobalVar 'correlationId')}}
{
 "expTime": "2020-05-05",
 "replayTo": "{{body 'replayTo' 'default'}}",
 "correlationId": "{{ getGlobalVar 'anotherId' }}"
}`
Callback definition:
`{
    "name": "It's callback",
   "correlationId": "{{ getGlobalVar 'anotherId' }}", 
    "returnUrl": "{{body 'replayTo' 'default'}}",
}`

**Reply by Matys-check** [reply, marked as answer]
Okay, it seems that after updating to the latest version (9.0), everything works as expected, so we can consider this solution finalized.
